bin/monitor.ml

```diff
@@ -0,0 +1,197 @@
+open Import
+open Fiber.O
+module Client = Dune_rpc_client.Client
+
+let poll_exn ?id client sub =
+  let+ poller = Client.poll ?id client sub in
+  match poller with
+  | Ok poller -> poller
+  | Error e -> raise (Dune_rpc_private.Version_error.E e)
+
+module Event = struct
+  (** Events that the render loop will process. *)
+
+  type t =
+    | Diagnostics of Dune_rpc.Diagnostic.Event.t list
+    | Progress of Dune_rpc.Progress.t
+    | Jobs of Dune_rpc.Job.Event.t list
+    | Poller_disconnected
+
+  let diagnostics l = Diagnostics l
+
+  let progress p = Progress p
+
+  let jobs l = Jobs l
+end
+
+module State = struct
+  (** Internal state of the render loop. *)
+
+  type t =
+    { mutable quit : bool
```

Comment:
Why do you need this flag? Isn't having the event in the mvar enough?

Reply:
If we don't finish reading all the events, the other loops will get stuck trying to write their events.

Reply:
Okay, I see. In that case, you should also make sure you aren't sending events past when `quit` is set in the `fetch_loop`.

Reply:
I don't have access to the state in the fetch loops so I will have to maybe do something else. There haven't been any issues with this from my testing however.

Reply:
Good to hear that it's not an issue, but it's still confusing to write events after quit.

You can change:

```
; mutable quit : bool
```

to:

```
; quit : bool ref
```

and then just share the quit reference with the fetch loops.